File: server/src/api/memory.ts
import { Hono } from "hono";
import { z } from "zod";
import { zValidator } from "@hono/zod-validator";
import {
  addEntity,
  addRelationship,
  getCommunities,
  getGraph,
  queryMemory,
  removeEntity,
} from "../services/memory.js";
import { listEntities } from "../services/graph.js";

export const memoryRouter = new Hono();

// --- Schemas ---
const querySchema = z.object({
  query: z.string().min(1),
  mode: z.enum(["auto", "local", "global", "drift", "basic"]).default("auto"),
  limit: z.number().int().positive().default(10),
  time_range: z
    .object({
      start: z.string().datetime().optional(),
      end: z.string().datetime().optional(),
    })
    .optional(),
  entity_filter: z.array(z.string()).optional(),
});

const entitySchema = z.object({
  name: z.string().min(1),
  type: z.string().min(1),
  properties: z.record(z.unknown()).optional(),
});

const relationshipSchema = z.object({
  source: z.string().min(1),
  target: z.string().min(1),
  type: z.string().min(1),
  properties: z.record(z.unknown()).optional(),
});

// --- POST /v1/memory/query ---
memoryRouter.post("/query", zValidator("json", querySchema), async (c) => {
  const body = c.req.valid("json");
  return c.json(
    await queryMemory({
      query: body.query,
      mode: body.mode,
      limit: body.limit,
      entityFilter: body.entity_filter,
    }),
  );
});

// --- GET /v1/memory/entities ---
memoryRouter.get("/entities", async (c) => {
  const type = c.req.query("type");
  const search = c.req.query("search");
  const limit = parseInt(c.req.query("limit") ?? "50", 10);
  const result = await listEntities({ type, search, limit });
  return c.json({
    entities: result.entities.map((entity) => ({
      id: entity.id,
      name: entity.name,
      type: entity.type,
      properties: entity.properties,
      created_at: entity.createdAt,
    })),
    total: result.total,
    limit,
  });
});

// --- POST /v1/memory/entities ---
memoryRouter.post("/entities", zValidator("json", entitySchema), async (c) => {
  const body = c.req.valid("json");
  return c.json(await addEntity(body), 201);
});

// --- DELETE /v1/memory/entities/:id ---
memoryRouter.delete("/entities/:id", async (c) => {
  const id = String(c.req.param("id"));
  return c.json(await removeEntity(id));
});

// --- POST /v1/memory/relationships ---
memoryRouter.post(
  "/relationships",
  zValidator("json", relationshipSchema),
  async (c) => {
    const body = c.req.valid("json");
    return c.json(await addRelationship(body), 201);
  },
);

// --- GET /v1/memory/communities ---
memoryRouter.get("/communities", async (c) => {
  const level = parseInt(c.req.query("level") ?? "0", 10);
  return c.json(await getCommunities(level));
});

// --- GET /v1/memory/graph ---
memoryRouter.get("/graph", async (c) => {
  const depth = parseInt(c.req.query("depth") ?? "2", 10);
  const center = c.req.query("center");
  return c.json(await getGraph(center, depth));
});

File: sdk/python/kraken/memory.py
"""GraphRAG memory operations."""

from __future__ import annotations

from typing import Any, Literal

from kraken._transport import Transport
from kraken.models import Entity, GraphView, MemoryQueryResult, Relationship

QueryMode = Literal["auto", "local", "global", "drift", "basic"]


class Memory:
    """Query and manage the agent's GraphRAG knowledge graph."""

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    # --- Query ---

    def query(
        self,
        query: str,
        *,
        mode: QueryMode = "auto",
        limit: int = 10,
        time_start: str | None = None,
        time_end: str | None = None,
        entity_filter: list[str] | None = None,
    ) -> MemoryQueryResult:
        """
        Query the agent's memory using GraphRAG multi-mode search.

        Args:
            query: Natural language question.
            mode: Search strategy — "auto" lets the agent pick.
            limit: Max results.
            time_start: ISO datetime lower bound.
            time_end: ISO datetime upper bound.
            entity_filter: Only search within these entity IDs.
        """
        payload: dict[str, Any] = {"query": query, "mode": mode, "limit": limit}
        if time_start or time_end:
            payload["time_range"] = {}
            if time_start:
                payload["time_range"]["start"] = time_start
            if time_end:
                payload["time_range"]["end"] = time_end
        if entity_filter:
            payload["entity_filter"] = entity_filter

        data = self._t.post("/v1/memory/query", json=payload)
        return MemoryQueryResult.model_validate(data)

    # --- Entities ---

    def list_entities(
        self,
        *,
        type: str | None = None,
        search: str | None = None,
        limit: int = 50,
    ) -> list[Entity]:
        """List entities in the knowledge graph."""
        params: dict[str, Any] = {"limit": limit}
        if type:
            params["type"] = type
        if search:
            params["search"] = search
        data = self._t.get("/v1/memory/entities", params=params)
        return [Entity.model_validate(e) for e in data["entities"]]

    def add_entity(
        self,
        name: str,
        type: str,
        properties: dict[str, Any] | None = None,
    ) -> Entity:
        """Add an entity to the knowledge graph."""
        payload: dict[str, Any] = {"name": name, "type": type}
        if properties:
            payload["properties"] = properties
        data = self._t.post("/v1/memory/entities", json=payload)
        return Entity.model_validate(data)

    def delete_entity(self, entity_id: str) -> None:
        """Remove an entity from the knowledge graph."""
        self._t.delete(f"/v1/memory/entities/{entity_id}")

    # --- Relationships ---

    def add_relationship(
        self,
        source: str,
        target: str,
        type: str,
        properties: dict[str, Any] | None = None,
    ) -> Relationship:
        """Add a relationship between two entities."""
        payload: dict[str, Any] = {"source": source, "target": target, "type": type}
        if properties:
            payload["properties"] = properties
        data = self._t.post("/v1/memory/relationships", json=payload)
        return Relationship.model_validate(data)

    # --- Graph ---

    def graph(self, *, center: str | None = None, depth: int = 2) -> GraphView:
        """Get a subgraph for visualization or traversal."""
        params: dict[str, Any] = {"depth": depth}
        if center:
            params["center"] = center
        data = self._t.get("/v1/memory/graph", params=params)
        return GraphView.model_validate(data)
